From 51f9b1771f14b71b0ef8184e61bced46370540da Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Tue, 30 Mar 2021 10:54:54 -0500
Subject: [PATCH 063/196] Add BIP341 sighash support to bitcoin-tx (#116)

Now sign=x|BIP341|FORKID and sign=x|BIP341|FORKID|ANYONECANPAY can be used.

Note: The BIP341 signatures are still invalid due to SigHashType::isDefined returning false, therefore the tests have signatures ending in 6x or ex instead of [x|BIP341|FORKID] etc.
This will need to be changed once SigHashType::isDefined allows BIP341 sigs.
---
 src/bitcoin-tx.cpp                            | 11 ++-
 src/script/sign.cpp                           |  4 +-
 src/script/sign.h                             |  1 +
 test/util/data/bitcoin-util-test.json         | 84 +++++++++++++++++++
 test/util/data/txcreatesignbip341v1-all.json  | 34 ++++++++
 ...txcreatesignbip341v2-all-anyonecanpay.json | 34 ++++++++
 test/util/data/txcreatesignbip341v2-all.json  | 34 ++++++++
 ...xcreatesignbip341v2-none-anyonecanpay.json | 34 ++++++++
 test/util/data/txcreatesignbip341v2-none.json | 34 ++++++++
 ...reatesignbip341v2-single-anyonecanpay.json | 34 ++++++++
 .../data/txcreatesignbip341v2-single.json     | 34 ++++++++
 11 files changed, 335 insertions(+), 3 deletions(-)
 create mode 100644 test/util/data/txcreatesignbip341v1-all.json
 create mode 100644 test/util/data/txcreatesignbip341v2-all-anyonecanpay.json
 create mode 100644 test/util/data/txcreatesignbip341v2-all.json
 create mode 100644 test/util/data/txcreatesignbip341v2-none-anyonecanpay.json
 create mode 100644 test/util/data/txcreatesignbip341v2-none.json
 create mode 100644 test/util/data/txcreatesignbip341v2-single-anyonecanpay.json
 create mode 100644 test/util/data/txcreatesignbip341v2-single.json

diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index ed10a029d..ce9beb119 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -523,7 +523,7 @@ static void MutateTxDelOutput(CMutableTransaction &tx,
     tx.vout.erase(tx.vout.begin() + outIdx);
 }
 
-static const unsigned int N_SIGHASH_OPTS = 12;
+static const unsigned int N_SIGHASH_OPTS = 18;
 static const struct {
     const char *flagStr;
     int flags;
@@ -543,6 +543,15 @@ static const struct {
      SIGHASH_NONE | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
     {"SINGLE|FORKID|ANYONECANPAY",
      SIGHASH_SINGLE | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
+    {"ALL|BIP341|FORKID", SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_FORKID},
+    {"NONE|BIP341|FORKID", SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_FORKID},
+    {"SINGLE|BIP341|FORKID", SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_FORKID},
+    {"ALL|BIP341|FORKID|ANYONECANPAY",
+     SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
+    {"NONE|BIP341|FORKID|ANYONECANPAY",
+     SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
+    {"SINGLE|BIP341|FORKID|ANYONECANPAY",
+     SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
 };
 
 static bool findSigHashFlags(SigHashType &sigHashType,
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 270e3922c..4985980a7 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -19,7 +19,7 @@ MutableTransactionSignatureCreator::MutableTransactionSignatureCreator(
     const Amount &amountIn, SigHashType sigHashTypeIn,
     const PrecomputedTransactionData &txdata)
     : txTo(txToIn), nIn(nInIn), amount(amountIn), sigHashType(sigHashTypeIn),
-      checker(txTo, nIn, amountIn, txdata) {}
+      checker(txTo, nIn, amountIn, txdata), m_txdata(txdata) {}
 
 bool MutableTransactionSignatureCreator::CreateSig(
     const SigningProvider &provider, std::vector<uint8_t> &vchSig,
@@ -31,7 +31,7 @@ bool MutableTransactionSignatureCreator::CreateSig(
     }
     uint256 hash;
     if (!SignatureHash(hash, execdata, scriptCode, *txTo, nIn, sigHashType,
-                       amount)) {
+                       amount, &m_txdata)) {
         return false;
     }
     if (!key.SignECDSA(hash, vchSig)) {
diff --git a/src/script/sign.h b/src/script/sign.h
index 6692983a3..9c93a4efa 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -42,6 +42,7 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
     Amount amount;
     SigHashType sigHashType;
     const MutableTransactionSignatureChecker checker;
+    const PrecomputedTransactionData &m_txdata;
 
 public:
     MutableTransactionSignatureCreator(
diff --git a/test/util/data/bitcoin-util-test.json b/test/util/data/bitcoin-util-test.json
index 75aebe2f5..66aaf4758 100644
--- a/test/util/data/bitcoin-util-test.json
+++ b/test/util/data/bitcoin-util-test.json
@@ -362,6 +362,90 @@
     "error_txt": "error: txid must be hexadecimal string (not '4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc5948512')",
     "description": "Tests the check for invalid txid valid hex, but too long"
   },
+  { "exec": "./bitcoin-tx",
+    "args":
+    ["-json",
+     "-create", "nversion=1",
+     "in=4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485:0",
+     "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
+     "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
+     "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
+     "sign=ALL|BIP341|FORKID"],
+    "output_cmp": "txcreatesignbip341v1-all.json",
+    "description": "Creates a new v1 transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash (output in json)"
+  },
+  { "exec": "./bitcoin-tx",
+    "args":
+    ["-json",
+     "-create",
+     "in=4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485:0",
+     "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
+     "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
+     "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
+     "sign=ALL|BIP341|FORKID|ANYONECANPAY"],
+    "output_cmp": "txcreatesignbip341v2-all-anyonecanpay.json",
+    "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with ANYONECANPAY (output in json)"
+  },
+  { "exec": "./bitcoin-tx",
+    "args":
+    ["-json",
+     "-create",
+     "in=4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485:0",
+     "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
+     "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
+     "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
+     "sign=ALL|BIP341|FORKID"],
+    "output_cmp": "txcreatesignbip341v2-all.json",
+    "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash (output in json)"
+  },
+  { "exec": "./bitcoin-tx",
+    "args":
+    ["-json",
+     "-create",
+     "in=4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485:0",
+     "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
+     "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
+     "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
+     "sign=NONE|BIP341|FORKID|ANYONECANPAY"],
+    "output_cmp": "txcreatesignbip341v2-none-anyonecanpay.json",
+    "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with NONE and ANYONECANPAY (output in json)"
+  },
+  { "exec": "./bitcoin-tx",
+    "args":
+    ["-json",
+     "-create",
+     "in=4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485:0",
+     "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
+     "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
+     "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
+     "sign=NONE|BIP341|FORKID"],
+    "output_cmp": "txcreatesignbip341v2-none.json",
+    "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with NONE (output in json)"
+  },
+  { "exec": "./bitcoin-tx",
+    "args":
+    ["-json",
+     "-create",
+     "in=4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485:0",
+     "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
+     "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
+     "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
+     "sign=SINGLE|BIP341|FORKID|ANYONECANPAY"],
+    "output_cmp": "txcreatesignbip341v2-single-anyonecanpay.json",
+    "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with SINGLE and ANYONECANPAY (output in json)"
+  },
+  { "exec": "./bitcoin-tx",
+    "args":
+    ["-json",
+     "-create",
+     "in=4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485:0",
+     "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
+     "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
+     "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
+     "sign=SINGLE|BIP341|FORKID"],
+    "output_cmp": "txcreatesignbip341v2-single.json",
+    "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with SINGLE (output in json)"
+  },
   { "exec": "./bitcoin-tx",
     "args":
     ["-create", "outpubkey=0:02a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397", "nversion=1"],
diff --git a/test/util/data/txcreatesignbip341v1-all.json b/test/util/data/txcreatesignbip341v1-all.json
new file mode 100644
index 000000000..4859fb06d
--- /dev/null
+++ b/test/util/data/txcreatesignbip341v1-all.json
@@ -0,0 +1,34 @@
+{
+    "txid": "19826ffef1f81f70b847b8eba42a45a4fa9d6e7cfa974c051a5215ac163c7e79",
+    "hash": "19826ffef1f81f70b847b8eba42a45a4fa9d6e7cfa974c051a5215ac163c7e79",
+    "version": 1,
+    "size": 191,
+    "locktime": 0,
+    "vin": [
+        {
+            "txid": "4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485",
+            "vout": 0,
+            "scriptSig": {
+                "asm": "304402204730710e773c6dccc773e473d1d67864c2489261fb810e79e7325ff42e1461a802202f6db04038b18d0077b911769cef7c30e405df971a2922f79d6d0118c98fcd5e61 02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
+                "hex": "47304402204730710e773c6dccc773e473d1d67864c2489261fb810e79e7325ff42e1461a802202f6db04038b18d0077b911769cef7c30e405df971a2922f79d6d0118c98fcd5e612102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+            },
+            "sequence": 4294967295
+        }
+    ],
+    "vout": [
+        {
+            "value": 0.00100000,
+            "n": 0,
+            "scriptPubKey": {
+                "asm": "OP_DUP OP_HASH160 06afd46bcdfd22ef94ac122aa11f241244a37ecc OP_EQUALVERIFY OP_CHECKSIG",
+                "hex": "76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac",
+                "reqSigs": 1,
+                "type": "pubkeyhash",
+                "addresses": [
+                    "1cMh228HTCiwS8ZsaakH8A8wze1JR5ZsP"
+                ]
+            }
+        }
+    ],
+    "hex": "01000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000006a47304402204730710e773c6dccc773e473d1d67864c2489261fb810e79e7325ff42e1461a802202f6db04038b18d0077b911769cef7c30e405df971a2922f79d6d0118c98fcd5e612102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5ffffffff01a0860100000000001976a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac00000000"
+}
diff --git a/test/util/data/txcreatesignbip341v2-all-anyonecanpay.json b/test/util/data/txcreatesignbip341v2-all-anyonecanpay.json
new file mode 100644
index 000000000..90aaf1e38
--- /dev/null
+++ b/test/util/data/txcreatesignbip341v2-all-anyonecanpay.json
@@ -0,0 +1,34 @@
+{
+    "txid": "c09531dce1eee038f0b5fdf47dd5d15a011159ef09240054000cb37ee0a85fd9",
+    "hash": "c09531dce1eee038f0b5fdf47dd5d15a011159ef09240054000cb37ee0a85fd9",
+    "version": 2,
+    "size": 191,
+    "locktime": 0,
+    "vin": [
+        {
+            "txid": "4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485",
+            "vout": 0,
+            "scriptSig": {
+                "asm": "3044022064bc2ead760a7d094a4fdd18a831434b5335163068afe03d369a513665079af10220425c1f9c3252f92a117a5aa33439cc45ab1326e2b2b7bc058c20d09cc59f6653e1 02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
+                "hex": "473044022064bc2ead760a7d094a4fdd18a831434b5335163068afe03d369a513665079af10220425c1f9c3252f92a117a5aa33439cc45ab1326e2b2b7bc058c20d09cc59f6653e12102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+            },
+            "sequence": 4294967295
+        }
+    ],
+    "vout": [
+        {
+            "value": 0.00100000,
+            "n": 0,
+            "scriptPubKey": {
+                "asm": "OP_DUP OP_HASH160 06afd46bcdfd22ef94ac122aa11f241244a37ecc OP_EQUALVERIFY OP_CHECKSIG",
+                "hex": "76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac",
+                "reqSigs": 1,
+                "type": "pubkeyhash",
+                "addresses": [
+                    "1cMh228HTCiwS8ZsaakH8A8wze1JR5ZsP"
+                ]
+            }
+        }
+    ],
+    "hex": "02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000006a473044022064bc2ead760a7d094a4fdd18a831434b5335163068afe03d369a513665079af10220425c1f9c3252f92a117a5aa33439cc45ab1326e2b2b7bc058c20d09cc59f6653e12102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5ffffffff01a0860100000000001976a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac00000000"
+}
diff --git a/test/util/data/txcreatesignbip341v2-all.json b/test/util/data/txcreatesignbip341v2-all.json
new file mode 100644
index 000000000..ab51e6ec7
--- /dev/null
+++ b/test/util/data/txcreatesignbip341v2-all.json
@@ -0,0 +1,34 @@
+{
+    "txid": "ddf1d3c333dbd1e95438a8839c38b37d7e5fabf86524e25cc5ebdb91fd15c50f",
+    "hash": "ddf1d3c333dbd1e95438a8839c38b37d7e5fabf86524e25cc5ebdb91fd15c50f",
+    "version": 2,
+    "size": 191,
+    "locktime": 0,
+    "vin": [
+        {
+            "txid": "4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485",
+            "vout": 0,
+            "scriptSig": {
+                "asm": "30440220700acff16c7fb62285d1bfb598fbb3ecfe3abfc9e94cb27d3e25f01bc419198d022064b9f58b3a0637f4673492b86a5b5aa78da87339ebf6826558f6abda7c557bf961 02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
+                "hex": "4730440220700acff16c7fb62285d1bfb598fbb3ecfe3abfc9e94cb27d3e25f01bc419198d022064b9f58b3a0637f4673492b86a5b5aa78da87339ebf6826558f6abda7c557bf9612102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+            },
+            "sequence": 4294967295
+        }
+    ],
+    "vout": [
+        {
+            "value": 0.00100000,
+            "n": 0,
+            "scriptPubKey": {
+                "asm": "OP_DUP OP_HASH160 06afd46bcdfd22ef94ac122aa11f241244a37ecc OP_EQUALVERIFY OP_CHECKSIG",
+                "hex": "76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac",
+                "reqSigs": 1,
+                "type": "pubkeyhash",
+                "addresses": [
+                    "1cMh228HTCiwS8ZsaakH8A8wze1JR5ZsP"
+                ]
+            }
+        }
+    ],
+    "hex": "02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000006a4730440220700acff16c7fb62285d1bfb598fbb3ecfe3abfc9e94cb27d3e25f01bc419198d022064b9f58b3a0637f4673492b86a5b5aa78da87339ebf6826558f6abda7c557bf9612102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5ffffffff01a0860100000000001976a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac00000000"
+}
diff --git a/test/util/data/txcreatesignbip341v2-none-anyonecanpay.json b/test/util/data/txcreatesignbip341v2-none-anyonecanpay.json
new file mode 100644
index 000000000..300ffc626
--- /dev/null
+++ b/test/util/data/txcreatesignbip341v2-none-anyonecanpay.json
@@ -0,0 +1,34 @@
+{
+    "txid": "fc57c51c8b9166f48907ba34f3496811682276280cc1cd36bebd3fdb988fcfa0",
+    "hash": "fc57c51c8b9166f48907ba34f3496811682276280cc1cd36bebd3fdb988fcfa0",
+    "version": 2,
+    "size": 191,
+    "locktime": 0,
+    "vin": [
+        {
+            "txid": "4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485",
+            "vout": 0,
+            "scriptSig": {
+                "asm": "304402201a964923f1cee629c5db30fefbf39a90ff388a21fc912d87bf02d98cca7b7beb0220090a58e5242cfa71992cc4c2c0ba49727c8bd0ab749c5789f7310c121af244f6e2 02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
+                "hex": "47304402201a964923f1cee629c5db30fefbf39a90ff388a21fc912d87bf02d98cca7b7beb0220090a58e5242cfa71992cc4c2c0ba49727c8bd0ab749c5789f7310c121af244f6e22102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+            },
+            "sequence": 4294967295
+        }
+    ],
+    "vout": [
+        {
+            "value": 0.00100000,
+            "n": 0,
+            "scriptPubKey": {
+                "asm": "OP_DUP OP_HASH160 06afd46bcdfd22ef94ac122aa11f241244a37ecc OP_EQUALVERIFY OP_CHECKSIG",
+                "hex": "76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac",
+                "reqSigs": 1,
+                "type": "pubkeyhash",
+                "addresses": [
+                    "1cMh228HTCiwS8ZsaakH8A8wze1JR5ZsP"
+                ]
+            }
+        }
+    ],
+    "hex": "02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000006a47304402201a964923f1cee629c5db30fefbf39a90ff388a21fc912d87bf02d98cca7b7beb0220090a58e5242cfa71992cc4c2c0ba49727c8bd0ab749c5789f7310c121af244f6e22102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5ffffffff01a0860100000000001976a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac00000000"
+}
diff --git a/test/util/data/txcreatesignbip341v2-none.json b/test/util/data/txcreatesignbip341v2-none.json
new file mode 100644
index 000000000..74d169bf5
--- /dev/null
+++ b/test/util/data/txcreatesignbip341v2-none.json
@@ -0,0 +1,34 @@
+{
+    "txid": "58760e36e4c244f169cd67d53b3ad7198f69227fed1beaa005648512a4351db1",
+    "hash": "58760e36e4c244f169cd67d53b3ad7198f69227fed1beaa005648512a4351db1",
+    "version": 2,
+    "size": 191,
+    "locktime": 0,
+    "vin": [
+        {
+            "txid": "4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485",
+            "vout": 0,
+            "scriptSig": {
+                "asm": "3044022026eff7ec933e50937d30cc8a7a4322db3c6f61c82afa47601e2c8aedebf86f5202202c173e2aae3bac8359196b6067ce407875913fe72bc91ef4337f8ee22cad140f62 02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
+                "hex": "473044022026eff7ec933e50937d30cc8a7a4322db3c6f61c82afa47601e2c8aedebf86f5202202c173e2aae3bac8359196b6067ce407875913fe72bc91ef4337f8ee22cad140f622102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+            },
+            "sequence": 4294967295
+        }
+    ],
+    "vout": [
+        {
+            "value": 0.00100000,
+            "n": 0,
+            "scriptPubKey": {
+                "asm": "OP_DUP OP_HASH160 06afd46bcdfd22ef94ac122aa11f241244a37ecc OP_EQUALVERIFY OP_CHECKSIG",
+                "hex": "76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac",
+                "reqSigs": 1,
+                "type": "pubkeyhash",
+                "addresses": [
+                    "1cMh228HTCiwS8ZsaakH8A8wze1JR5ZsP"
+                ]
+            }
+        }
+    ],
+    "hex": "02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000006a473044022026eff7ec933e50937d30cc8a7a4322db3c6f61c82afa47601e2c8aedebf86f5202202c173e2aae3bac8359196b6067ce407875913fe72bc91ef4337f8ee22cad140f622102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5ffffffff01a0860100000000001976a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac00000000"
+}
diff --git a/test/util/data/txcreatesignbip341v2-single-anyonecanpay.json b/test/util/data/txcreatesignbip341v2-single-anyonecanpay.json
new file mode 100644
index 000000000..4e36a60af
--- /dev/null
+++ b/test/util/data/txcreatesignbip341v2-single-anyonecanpay.json
@@ -0,0 +1,34 @@
+{
+    "txid": "992e7ac4ebba029956f6df08817de49a6ccd7b296d6f68bf8b7a6846fab82eaf",
+    "hash": "992e7ac4ebba029956f6df08817de49a6ccd7b296d6f68bf8b7a6846fab82eaf",
+    "version": 2,
+    "size": 191,
+    "locktime": 0,
+    "vin": [
+        {
+            "txid": "4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485",
+            "vout": 0,
+            "scriptSig": {
+                "asm": "3044022077f44e39e3a3ae076341efae954f84a10211576fffc0106e20d979bcb8aeb665022026f765d100c4b3f0e1f3482a676fd60f725d81d5afd2172adfa23281bbdff97de3 02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
+                "hex": "473044022077f44e39e3a3ae076341efae954f84a10211576fffc0106e20d979bcb8aeb665022026f765d100c4b3f0e1f3482a676fd60f725d81d5afd2172adfa23281bbdff97de32102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+            },
+            "sequence": 4294967295
+        }
+    ],
+    "vout": [
+        {
+            "value": 0.00100000,
+            "n": 0,
+            "scriptPubKey": {
+                "asm": "OP_DUP OP_HASH160 06afd46bcdfd22ef94ac122aa11f241244a37ecc OP_EQUALVERIFY OP_CHECKSIG",
+                "hex": "76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac",
+                "reqSigs": 1,
+                "type": "pubkeyhash",
+                "addresses": [
+                    "1cMh228HTCiwS8ZsaakH8A8wze1JR5ZsP"
+                ]
+            }
+        }
+    ],
+    "hex": "02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000006a473044022077f44e39e3a3ae076341efae954f84a10211576fffc0106e20d979bcb8aeb665022026f765d100c4b3f0e1f3482a676fd60f725d81d5afd2172adfa23281bbdff97de32102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5ffffffff01a0860100000000001976a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac00000000"
+}
diff --git a/test/util/data/txcreatesignbip341v2-single.json b/test/util/data/txcreatesignbip341v2-single.json
new file mode 100644
index 000000000..68134682d
--- /dev/null
+++ b/test/util/data/txcreatesignbip341v2-single.json
@@ -0,0 +1,34 @@
+{
+    "txid": "334256a59387797146757632b4c33bcfde0152b340e9a7639513cc25de6c32bc",
+    "hash": "334256a59387797146757632b4c33bcfde0152b340e9a7639513cc25de6c32bc",
+    "version": 2,
+    "size": 190,
+    "locktime": 0,
+    "vin": [
+        {
+            "txid": "4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485",
+            "vout": 0,
+            "scriptSig": {
+                "asm": "3043021f302b36058b8617d81259d9da25e91b321eecf9d3826f36b83f7671572d0aa2022049b222b2e9d8356d22efcda07931d9360e2a4cdadabb62723463b5111d4a59ae63 02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5",
+                "hex": "463043021f302b36058b8617d81259d9da25e91b321eecf9d3826f36b83f7671572d0aa2022049b222b2e9d8356d22efcda07931d9360e2a4cdadabb62723463b5111d4a59ae632102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+            },
+            "sequence": 4294967295
+        }
+    ],
+    "vout": [
+        {
+            "value": 0.00100000,
+            "n": 0,
+            "scriptPubKey": {
+                "asm": "OP_DUP OP_HASH160 06afd46bcdfd22ef94ac122aa11f241244a37ecc OP_EQUALVERIFY OP_CHECKSIG",
+                "hex": "76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac",
+                "reqSigs": 1,
+                "type": "pubkeyhash",
+                "addresses": [
+                    "1cMh228HTCiwS8ZsaakH8A8wze1JR5ZsP"
+                ]
+            }
+        }
+    ],
+    "hex": "02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d0000000069463043021f302b36058b8617d81259d9da25e91b321eecf9d3826f36b83f7671572d0aa2022049b222b2e9d8356d22efcda07931d9360e2a4cdadabb62723463b5111d4a59ae632102c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5ffffffff01a0860100000000001976a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac00000000"
+}
-- 
2.39.5 (Apple Git-154)

